Library/Formula/ghc.rb

```diff
@@ -10,163 +12,91 @@ class Ghc < Formula
+                n <- get :: Get Int getMany n
+      EOS
     end
   end
 
-  stable do
-    resource "testsuite" do
-      url "https://downloads.haskell.org/~ghc/7.10.1/ghc-7.10.1-testsuite.tar.xz"
-      sha256 "33bbdfcfa50363526ea9671c8c1f01b7c5dec01372604d45cbb53bb2515298cb"
+  resource "binary" do
+    if MacOS.version <= :lion
+      url "https://downloads.haskell.org/~ghc/7.6.3/ghc-7.6.3-x86_64-apple-darwin.tar.bz2"
+      sha256 "f7a35bea69b6cae798c5f603471a53b43c4cc5feeeeb71733815db6e0a280945"
+    else
+      url "https://downloads.haskell.org/~ghc/7.10.1/ghc-7.10.1-x86_64-apple-darwin.tar.xz"
+      sha256 "bc45de19efc831f7d5a3fe608ba4ebcd24cc0f414cac4bc40ef88a04640583f6"
     end
   end
 
-  fails_with :llvm do
-    cause <<-EOS.undent
-      cc1: error: unrecognized command line option "-Wno-invalid-pp-token"
-      cc1: error: unrecognized command line option "-Wno-unicode"
-    EOS
-  end
-
-  if build.build_32_bit? || !MacOS.prefer_64_bit? || MacOS.version < :mavericks
-    fails_with :clang do
-      cause <<-EOS.undent
-        Building with Clang configures GHC to use Clang as its preprocessor,
-        which causes subsequent GHC-based builds to fail.
-      EOS
-    end
+  resource "testsuite" do
+    url "https://downloads.haskell.org/~ghc/7.10.1/ghc-7.10.1-testsuite.tar.xz"
+    sha256 "33bbdfcfa50363526ea9671c8c1f01b7c5dec01372604d45cbb53bb2515298cb"
   end
 
   def install
-    ENV.m32 if build.build_32_bit?
-
-    # Build a static gmp (to avoid dynamic linking to ghc)
-    gmp_prefix = buildpath/"gmp-static"
-    resource("gmp").stage do
-      gmp_args = ["--prefix=#{gmp_prefix}", "--enable-cxx", "--enable-shared=no"]
-      gmp_args << "ABI=32" if build.build_32_bit?
+    # Build a static gmp rather than in-tree gmp, otherwise it links to brew's.
+    gmp = libexec/"integer-gmp"
 
-      # https://github.com/Homebrew/homebrew/issues/20693
-      gmp_args << "--disable-assembly" if build.build_32_bit? || build.bottle?
+    # MPN_PATH: The lowest common denomenator asm paths that work on Darwin,
+    # corresponding to Yonah and Merom. Obviates --disable-assembly.
+    ENV["MPN_PATH"] = "x86_64/fastsse x86_64/core2 x86_64 generic"
 
-      system "./configure", *gmp_args
+    # GMP *does not* build PIC by default without shared libraries so --with-pic
+    # is mandatory or else you'll get "illegal text relocs" errors.
+    resource("gmp").stage do
+      system "./configure", "--prefix=#{gmp}", "--with-pic", "--disable-shared"
       system "make"
       system "make", "check"
-      ENV.deparallelize
-      system "make", "install"
+      ENV.deparallelize { system "make", "install" }
     end
 
-    # Move the main tarball contents into a subdirectory
-    (buildpath+"Ghcsource").install Dir["*"]
+    args = ["--with-gmp-includes=#{gmp}/include",
+            "--with-gmp-libraries=#{gmp}/lib",
+            "--with-ld=ld", # Avoid hardcoding superenv's ld on older machines.
```

Comment:
Because otherwise it will not be supplied the right arguments after being filtered through either `collect2`, or the ENV wrapper. This was the cause of failure on #39025.